src/reactive-membrane.ts

```diff
@@ -76,31 +90,150 @@ const defaultValueMutated: ReactiveMembraneMutationCallback = (obj: any, key: Pr
-            };
+            descriptor.get = reserveGetterMap.get(get) || get;
         }
         if (!isUndefined(set)) {
-            descriptor.set = function(value) {
-                // At this point we don't have a clear indication of whether
-                // or not a valid mutation will occur, we don't have the key,
-                // and we are not sure why and how they are invoking this setter.
-                // Nevertheless we preserve the original semantics by invoking the
-                // original setter with the original target and the unwrapped value
-                set.call(unwrap(this), membrane.unwrapProxy(value));
-            };
+            descriptor.set = reverseSetterMap.get(set) || set;
         }
     }
     return descriptor;
 }
 
+export function copyDescriptorIntoShadowTarget(
+    handler: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget,
+    key: PropertyKey,
+) {
+    const { originalTarget } = handler;
+    // Note: a property might get defined multiple times in the shadowTarget
+    //       but it will always be compatible with the previous descriptor
+    //       to preserve the object invariants, which makes these lines safe.
+    const normalizedDescriptor = getOwnPropertyDescriptor(originalTarget, key);
+    if (!isUndefined(normalizedDescriptor)) {
+        const blueDesc = wrapDescriptor(handler, normalizedDescriptor);
+        ObjectDefineProperty(shadowTarget, key, blueDesc);
+    }
+}
+
+export function lockShadowTarget(
+    handler: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget
+): void {
+    const { originalTarget } = handler;
+    const targetKeys = ArrayConcat.call(getOwnPropertyNames(originalTarget), getOwnPropertySymbols(originalTarget));
+    targetKeys.forEach((key: PropertyKey) => {
+        copyDescriptorIntoShadowTarget(handler, shadowTarget, key);
+    });
+
+    preventExtensions(shadowTarget);
+}
+
+export function isExtensibleMembraneTrap(
+    this: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget
+): boolean {
+    const { originalTarget } = this;
+    // optimization to avoid attempting to lock down the shadowTarget multiple times
+    if (!isExtensible(shadowTarget)) {
+        return false; // was already locked down
+    }
+    if (!isExtensible(originalTarget)) {
+        lockShadowTarget(this, shadowTarget);
+        return false;
+    }
+    return true;
+}
+
+export function getOwnPropertyDescriptorMembraneTrap(
+    this: MembraneProxyHandler,
+    shadowTarget: ReactiveMembraneShadowTarget,
+    key: PropertyKey
+): PropertyDescriptor | undefined {
+    const { originalTarget, membrane: { valueObserved } } = this;
+
+    // keys looked up via hasOwnProperty need to be reactive
+    valueObserved(originalTarget, key);
+
+    const desc = getOwnPropertyDescriptor(originalTarget, key);
+    if (isUndefined(desc)) {
+        return desc;
+    }
+
+    if (desc.configurable === false) {
```

Comment:
this is probably the block with the bigger changes... in the previous version of this, from each handler, the descriptor from the shadowTarget was read, and used when possible... that on itself poses many problems, but the most important one is the following:

* you might get an old value from an old installed non-configurable descriptor that is writable, e.g.: array.length

* [x] need tests